### PR TITLE
Elfobj: Simplify hashtable initialization with static foreach

### PR DESCRIPTION
**Pull request by Geod24**
Remove some empty functions, simplify repetitive code and improve scoping.

### CODE REVIEW

File: src/dmd/backend/elfobj.d [hunk 1]
```diff
@@ -818,20 +762,8 @@ Obj Obj_init(Outbuffer *objbuf, const(char)* filename, const(char)* csegname)
...
-        namidx = NAMIDX_STRTAB;    *section_names_hashtable.get(namidx, namidx + cast(uint)strlen(section_names_init64.ptr + namidx)) = namidx;
-        namidx = NAMIDX_SYMTAB;    *section_names_hashtable.get(namidx, namidx + cast(uint)strlen(section_names_init64.ptr + namidx)) = namidx;
-        namidx = NAMIDX_SHSTRTAB;  *section_names_hashtable.get(namidx, namidx + cast(uint)strlen(section_names_init64.ptr + namidx)) = namidx;
-        namidx = NAMIDX_COMMENT;   *section_names_hashtable.get(namidx, namidx + cast(uint)strlen(section_names_init64.ptr + namidx)) = namidx;
-        namidx = NAMIDX_NOTE;      *section_names_hashtable.get(namidx, namidx + cast(uint)strlen(section_names_init64.ptr + namidx)) = namidx;
-        namidx = NAMIDX_GNUSTACK;  *section_names_hashtable.get(namidx, namidx + cast(uint)strlen(section_names_init64.ptr + namidx)) = namidx;
-        namidx = NAMIDX_CDATAREL;  *section_names_hashtable.get(namidx, namidx + cast(uint)strlen(section_names_init64.ptr + namidx)) = namidx;
+        static foreach (idx; __traits(allMembers, NAMIDX)[1 .. $])
```
Comment (thewilsonator) [inline src/dmd/backend/elfobj.d]: Do we use `static foreach` elsewhere in the compiler?

File: src/dmd/backend/elfobj.d [hunk 2]
```diff
@@ -804,34 +748,22 @@ Obj Obj_init(Outbuffer *objbuf, const(char)* filename, const(char)* csegname)
...
+        elf_newsection2(NAMIDX.STRTAB,SHT_STRTAB, 0,                    0,0,0,0,0, 1,0);
+        elf_newsection2(NAMIDX.SYMTAB,SHT_SYMTAB, 0,                    0,0,0,0,0, 8,0);
+        elf_newsection2(NAMIDX.SHSTRTAB,SHT_STRTAB, 0,                  0,0,0,0,0, 1,0);
+        elf_newsection2(NAMIDX.COMMENT, SHT_PROGBITS,0,                 0,0,0,0,0, 1,0);
+        elf_newsection2(NAMIDX.NOTE,SHT_NOTE,   0,                      0,0,0,0,0, 1,0);
+        elf_newsection2(NAMIDX.GNUSTACK,SHT_PROGBITS,0,                 0,0,0,0,0, 1,0);
+        elf_newsection2(NAMIDX.CDATAREL,SHT_PROGBITS,SHF_ALLOC|SHF_WRITE,0,0,0,0,0, 16,0);
+
+        static foreach (idx; __traits(allMembers, NAMIDX)[1 .. $])
+            *section_names_hashtable.get(idx, cast(uint)section_names_init64.sizeof) = idx;
```
Comment (thewilsonator) [inline src/dmd/backend/elfobj.d]:
```
*section_names_hashtable.get(mixin("NAMIDX"~idx), cast(uint)section_names_init64.sizeof) = idx;
```

File: src/dmd/backend/elfobj.d [hunk 3]
```diff
@@ -859,34 +791,22 @@ Obj Obj_init(Outbuffer *objbuf, const(char)* filename, const(char)* csegname)
...
+        elf_newsection2(NAMIDX.STRTAB,SHT_STRTAB, 0,                    0,0,0,0,0, 1,0);
+        elf_newsection2(NAMIDX.SYMTAB,SHT_SYMTAB, 0,                    0,0,0,0,0, 4,0);
+        elf_newsection2(NAMIDX.SHSTRTAB,SHT_STRTAB, 0,                  0,0,0,0,0, 1,0);
+        elf_newsection2(NAMIDX.COMMENT, SHT_PROGBITS,0,                 0,0,0,0,0, 1,0);
+        elf_newsection2(NAMIDX.NOTE,SHT_NOTE,   0,                      0,0,0,0,0, 1,0);
+        elf_newsection2(NAMIDX.GNUSTACK,SHT_PROGBITS,0,                 0,0,0,0,0, 1,0);
+        elf_newsection2(NAMIDX.CDATAREL,SHT_PROGBITS,SHF_ALLOC|SHF_WRITE,0,0,0,0,0, 1,0);
+
+        static foreach (idx; __traits(allMembers, NAMIDX)[1 .. $])
+            *section_names_hashtable.get(idx, cast(uint)section_names_init64.sizeof) = idx;
```
Comment (thewilsonator) [inline src/dmd/backend/elfobj.d]: Ditto, also the 64is unnecessary.

File: src/dmd/backend/elfobj.d (hunk 1 above)
Comment (Geod24) [reply src/dmd/backend/elfobj.d]: Good point, removed

File: src/dmd/backend/elfobj.d [hunk 4]
```diff
@@ -804,34 +748,25 @@ Obj Obj_init(Outbuffer *objbuf, const(char)* filename, const(char)* csegname)
...
+        elf_newsection2(NAMIDX.SYMTAB,SHT_SYMTAB, 0,                    0,0,0,0,0, 8,0);
+        elf_newsection2(NAMIDX.SHSTRTAB,SHT_STRTAB, 0,                  0,0,0,0,0, 1,0);
+        elf_newsection2(NAMIDX.COMMENT, SHT_PROGBITS,0,                 0,0,0,0,0, 1,0);
+        elf_newsection2(NAMIDX.NOTE,SHT_NOTE,   0,                      0,0,0,0,0, 1,0);
+        elf_newsection2(NAMIDX.GNUSTACK,SHT_PROGBITS,0,                 0,0,0,0,0, 1,0);
+        elf_newsection2(NAMIDX.CDATAREL,SHT_PROGBITS,SHF_ALLOC|SHF_WRITE,0,0,0,0,0, 16,0);
+
+        foreach (idxname; __traits(allMembers, NAMIDX)[1 .. $])
+        {
+            NAMIDX idx = mixin("NAMEIDX." ~ idxname);
```
Comment (thewilsonator) [inline src/dmd/backend/elfobj.d]: `mixin("NAMIDX."`, no `E`.

File: src/dmd/backend/elfobj.d [hunk 5]
```diff
@@ -859,34 +794,25 @@ Obj Obj_init(Outbuffer *objbuf, const(char)* filename, const(char)* csegname)
...
+        elf_newsection2(NAMIDX.SYMTAB,SHT_SYMTAB, 0,                    0,0,0,0,0, 4,0);
+        elf_newsection2(NAMIDX.SHSTRTAB,SHT_STRTAB, 0,                  0,0,0,0,0, 1,0);
+        elf_newsection2(NAMIDX.COMMENT, SHT_PROGBITS,0,                 0,0,0,0,0, 1,0);
+        elf_newsection2(NAMIDX.NOTE,SHT_NOTE,   0,                      0,0,0,0,0, 1,0);
+        elf_newsection2(NAMIDX.GNUSTACK,SHT_PROGBITS,0,                 0,0,0,0,0, 1,0);
+        elf_newsection2(NAMIDX.CDATAREL,SHT_PROGBITS,SHF_ALLOC|SHF_WRITE,0,0,0,0,0, 1,0);
+
+        foreach (idxname; __traits(allMembers, NAMIDX)[1 .. $])
+        {
+            NAMIDX idx = mixin("NAMEIDX." ~ idxname);
```
Comment (thewilsonator) [inline src/dmd/backend/elfobj.d]: Ditto.

**Comment by thewilsonator** [comment]
The only things different between the two branches are 

```diff
-static immutable char[106 + 12] section_names_init =
+static immutable char[107 + 12] section_names_init64 =
...
-section_names.reserve(1024);
+section_names.reserve(100*1024);
```
and a few of the constants  in the last two columns .

The first I presume is an oversight, the second I'm not so sure about.

I think it would be better to get rid of the branch completely.

**Comment by Geod24** [comment]
> The only things different between the two branches are

No ? The content of the array is different. The last section name is ".rel.data" for the 32 bits version, "rela.data" for the 64 bits version. I can have a look at factoring more code in the future, but the branch cannot be trivially eliminated.

**Comment by WalterBright** [comment]
Since the changes to these 3 files appear to be independent of each other, it would be better to make this 3 separate PRs, one for each file, to make it clear they are independent (and will make it 3x easier to digest).

**Comment by Geod24** [comment]
Done

**Review by thewilsonator — APPROVED** [review]

**Review by atilaneves — APPROVED** [review]

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**